src/OpenTelemetry/Metrics/HistogramBuckets.cs

```diff
@@ -36,9 +40,38 @@ public class HistogramBuckets
 
         internal int IsCriticalSectionOccupied = 0;
 
+        private readonly Bucket root;
+
+        private readonly Func<double, int> findHistogramBucketIndex;
+
         internal HistogramBuckets(double[] explicitBounds)
         {
             this.ExplicitBounds = explicitBounds;
+            this.findHistogramBucketIndex = this.FindBucketIndexLinear;
+            if (explicitBounds != null && explicitBounds.Length >= DefaultHistogramCountForBinarySearch)
+            {
+                this.root = ConstructBalancedTree(explicitBounds, 0, explicitBounds.Length);
+                this.findHistogramBucketIndex = this.FindBucketIndexBinary;
+
+                static Bucket ConstructBalancedTree(double[] values, int min, int max)
+                {
+                    if (min == max)
+                    {
+                        return null;
+                    }
+
+                    int median = min + ((max - min) / 2);
+                    return new Bucket
+                    {
+                        Index = median,
+                        UpperBoundInclusive = values[median],
+                        LowerBoundExclusive = median > 0 ? values[median - 1] : double.MinValue,
```

Comment:
Use `double.NegativeInfinity` instead of `double.MinValue`?

Reply:
Good call. I think `double.NegativeInfinity` would be fine here. @alanwest you agree?